### PR TITLE
Devise

### PR DESCRIPTION
**Pull request by EmiLoKiJu**
Remove current_user method in ApplicationController, devise will provide us with one. ✔️ 
[Install and setup devise](https://github.com/heartcombo/devise#getting-started). ✔️ 
Can register a new user. ✔️ 
User logs in with a combination of email and password. ✔️ 
Hashed password should be stored in the database. ✔️ 

Ask for confirmation of email. ✔️ 

Confirmation email:
image

When clicking the link that is shown in the vscode terminal, this is what happens:
image

Can reset password. ✔️ 

The mail you should receive when resetting the password:
image

Changing password:
image

Password changed:
image

[Modify the views in devise](https://github.com/heartcombo/devise#configuring-views) for registration and login to match [the wireframes from the Sneak Peek](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/sneak_peek.md#:~:text=For%20this%20project%20you%20will%20have%20full%20freedom%20in%20terms%20of%20visual%20design%20but%20you%20will%20need%20to%20keep%20the%20following%20wireframes%3A) and your styling. ✔️

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -1,6 +1,2 @@
 class ApplicationController < ActionController::Base
-  def current_user
-    User.first
-  end
-  helper_method :current_user
 end
```
Comment (abdulhamiid) [inline app/controllers/application_controller.rb]: [OPTIONAL] - Protecting Content in the Controller

Let’s say our website includes routes that should only be made available to users who are logged in. For instance, a user might need to be logged in to be able to view all users, create a post, view comments, etc. We could add this code to our application controller to protect our routes: 🤗🛠

```ruby
class ApplicationController < ActionController::Base
  before_action :authenticate_user!
end
```

👉 [Basic Authorization with Devise](https://www.learnhowtoprogram.com/ruby-and-rails/authentication-and-authorization/basic-authorization-with-devise)

File: app/models/user.rb [hunk 2]
```diff
@@ -1,4 +1,10 @@
 class User < ApplicationRecord
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :validatable,
+         :confirmable
```
Comment (abdulhamiid) [inline app/models/user.rb:6]: After `sign_up`, please let's ensure the confirmation of email before the new user can log in, according to the [project requirements](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/blog_app_devise.md)

Confirmable is responsible for verifying if an account is already confirmed to sign in and send emails with confirmation instructions. 😊

File: app/views/devise/registrations/new.html.erb [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    <%= f.label :password_confirmation %><br />
+    <%= f.password_field :password_confirmation, autocomplete: "new-password" %>
+  </div>
+
+  <div class="actions">
+    <%= f.submit "Sign up" %>
+  </div>
+<% end %>
+
+<%= render "devise/shared/links" %>
```
Comment (abdulhamiid) [inline app/views/devise/registrations/new.html.erb]: Your design
image

Please modify the view in the devise for registration with your styling (similar to the [login page](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/sneak_peek.md))  improving the UX design. 💪😀

image

**Review by abdulhamiid — CHANGES_REQUESTED** [review]
# Status: Required Changes :recycle:
Hi Team 👋, 
Alt Text
Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!
<FEEL FREE TO CHANGE THIS MESSAGE TO YOUR PERSONAL WORDS>

## Highlights 🌟

- ✔️ Descriptions are very descriptive
- ✔️ Build a web app that requires the user to log in.
- ✔️ Use devise gem for authentication.

## Required Changes ♻️
_Please kindly check the comments under the review._

## Optional suggestions 🤷‍♂️
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

## Important Notes 📝
- Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
- Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/views/devise/registrations/new.html.erb (hunk 3 above)
Comment (KDlamini) [reply app/views/devise/registrations/new.html.erb]: Well resolved ✔️

File: app/models/user.rb (hunk 2 above)
Comment (KDlamini) [reply app/models/user.rb:6]: Not resolved, the user is still logged in before confirmation. ✖️

File: app/models/user.rb [hunk 4]
```diff
@@ -1,4 +1,10 @@
 class User < ApplicationRecord
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :validatable,
```
Comment (KDlamini) [inline app/models/user.rb:5]: Hi @EmiLoKiJu

The user is still automatically logged in after signing up. As requested by the previous reviewer, the user must confirm the email first before they are logged in.

- Kindly consider changing the redirect path after signing up and ensure that the app redirects to the login page and the user can only log in after the confirmation link has been clicked.

**Review by KDlamini — CHANGES_REQUESTED** [review]
### Hi @EmiLoKiJu 🙋‍♀️,  

You have done an incredible job implementing some of the changes requested by the previous reviewer. 👏🏾👏🏾👏🏾  
However, there are still some changes that aren't addressed yet. You are almost there! 💪🏾 

## Highlights 🌞 

✔️  Good PR title and description
✔️  Devise has been installed and setup correctly
✔️  Devise is  :database_authenticatable, :registerable, :recoverable, :rememberable, :validatable, and :confirmable
✔️  The views in devise for the registration and login match the wireframes

## Required Changes ♻️
 
**Kindly look through the comments below 👇 for required changes**

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

N/A

Cheers 🥂  and Happy coding!!! 👯 
Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me @KDlamini in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/models/user.rb (hunk 4 above)
Comment (EmiLoKiJu) [reply app/models/user.rb:5]: image

I have checked my app and I'm not logged in after creatin a new account.

**Review by nicupop729 — COMMENTED** [review]
Hello @EmiLoKiJu  👋,

_This is an additional review as a result of the second opinion request._

Great job on reaching this milestone!
Firstly, I'd like to emphasize the importance of genuine collaboration between pair programming partners, rather than just facade collaboration.

Upon cloning, installing, and running the project, I observed that the user isn't logged in until the link in the terminal is clicked. Consequently, it means that non-logged in users can access the app. Ideally, we should avoid this because it deviates from the usual behavior. However, since the project requirements do not explicitly mention this aspect, the project is ready for approval.

As this was the only reason that your project was not approved it will be approved as soon as possible.
Please ask for another review in your Dashboard and the next Code Reviewer will mark it as approve right away :)

Keep rocking 🚀!

**Comment by EmiLoKiJu** [comment]
Hi @nicupop729 
Thx for the extra review ;)
I can still see the changes required in the dashboard. Is that normal?
image

**Comment by nicupop729** [comment]
Hi @EmiLoKiJu,

If I am not wrong (please correct me if I am), you still have a review attempt, no?
Please use it to ask for another review from your dashboard and the next reviewer will mark the project as approved right away (as it was mentioned in my previous review).

**Review by KanzaTahreem — APPROVED** [review]
# Approved 🙌

Hi @EmiLoKiJu 

Good job implementing the suggested changes 💪 **_Your project is COMPLETE! There is nothing else to say other than... it's time to merge it! 🚀 Congratulations!_**



## Highlights 🎯
There are some excellent points that you have made perfectly. Keep up the great work! 👏

-  ✅ Devise is configured
-  ✅ Using the right GitFlow 
-  ✅ Linters are working properly

## Optional suggestions ❕
N/A

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👨‍💻💻

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**